File: js/src/RunePackage.ts
import { RuneSpec, parse } from "./rspec";
import * as zipArchive from "jszip";
export class RunePackage
{
    public metadata: RuneSpec;
    public content: ArrayBuffer;
}

export async function unwrap(buffer: ArrayBuffer): Promise<RunePackage>
{
    const zip = new zipArchive();
    const pkg = new RunePackage();
    await zip.loadAsync(buffer).catch((e) => { throw new Error(`failed open package. [${e}]`) });
    pkg.content = buffer;
    const entity = await zip.files["target.rspec"].async("text");
    const spec = parse(entity);

    pkg.metadata = spec;

    return pkg;
}

File: csharp/Ancient.ProjectSystem/RunePackage.cs
namespace Ancient.ProjectSystem
{
    using System;
    using System.IO;
    using System.IO.Compression;
    using System.Text;
    using System.Threading;
    using System.Threading.Tasks;
    using exceptions;
    using Newtonsoft.Json;
    using NuGet.Versioning;

    public class RunePackage : IDisposable, IAsyncDisposable
    {
        public string ID { get; set; }
        public NuGetVersion Version { get; set; }
        public MemoryStream Content { get; set; }
        public RuneSpec Spec { get; set; }

        public void Dispose()
            => Content?.Dispose();
        public ValueTask DisposeAsync()
            => Content?.DisposeAsync() ?? new ValueTask(Task.CompletedTask);

        public static async Task<RunePackage> Unwrap(byte[] buffer, CancellationToken cancellationToken = default)
        {
            await using var mem = new MemoryStream(buffer);
            using var zip = new ZipArchive(mem, ZipArchiveMode.Read);

            var entity = zip.GetEntry("target.rspec");

            if (entity is null)
                throw new CorruptedPackageException($"Could not find target.rspec in package.");

            await using var point = entity.Open();
            await using var memory = new MemoryStream();
            await point.CopyToAsync(memory, cancellationToken);
            var result = Encoding.UTF8.GetString(memory.ToArray());

            var spec = JsonConvert.DeserializeObject<RuneSpec>(result);

            return new RunePackage
            {
                Content = new MemoryStream(buffer),
                ID = spec.ID,
                Version = spec.Version,
                Spec = spec
            };
        }
    }
}